### DISCUSSION TITLE
Wiring together arduino Mega and 8266 to remote (OTA)update of Mega2586 sketches

### DISCUSSION THREAD

**Discussion by maxi15865** [JAndrassy/ArduinoOTA#111] [Q&A]
Hi, I am quite new to Arduino and I am trying to build a stable project, for my blind girlfriend, that uses OTA upgrades.
I would like to use a 8266 nodemcu wired to a Mega 2586 so that the 8266 will handle the wifi, OTA and remote debug and let the Mega take care of the core of the code, (sensors, dfplayer, etc..). I have problems  to follow the instructions on main page of ArduinoOta, no idea where to start.
A basic example of wiring and code would be great.

**Comment by JAndrassy** [comment]
you can use my WiFiEspAT library to use esp8266 as WiFi adapter. See the [SerialWiFiOTA](https://github.com/jandrassy/ArduinoOTA/blob/master/examples/SerialWiFiOTA/SerialWiFiOTA.ino) example.
but how to wire esp8266 is out of scope even for the WiFiEspAT library's [README](https://github.com/jandrassy/WiFiEspAT/blob/master/README.md).
Before wiring it to Mega, upload Espressif AT firmware version 1.7.4 into the NodeMcu.
Nodemcu requires to wire Gnd, RX, TX (and 5 V only if USB is not connected) 
On Arduino Mega use Serial1 pins RX1/TX1. Wire Serial as TX to RX (transmit to receive).
It is recommended to use logic level conversion from 5 V to 3.3 V for RX pin of esp8266. 

btw: If you upload from Linux and you don't need a WiFi library in the sketch, then it is possible to flash the Mega with avrdude over the esp8266 without the ArduinoOTA and WiFiEspAT library.

**Reply by maxi15865** [reply]
Thank you very much for your help! I´m going tu study your information. I´m so happy. Have a nice day.

**Comment by maxi15865** [comment]
Hi, Jandrassy. I´m not sure what bin download to use with my 8266. I found this on Espressif homepage:
"ESP8266 NonOS AT Bin V1.7.4
ESP8266 NonOS AT Bin will not be updated any more. We recommend that you use ESP8266 IDF AT Bin for new designs."
Is this the correct one?
Thank you.

**Reply by JAndrassy** [reply]
they have 2.0 based on RTOS but it is still not stable. 1.7.4 is the best AT firmware for esp8266 for now.

**Comment by maxi15865** [comment]
Sorry to bother you again Jandrassy, but you can help, you are my digital God now :) I have problems to understand how to flash the 8266, It seems that I must buy a dedicated device but I found a page where is shown  how to flash the 8266 using an arduino mega. I already have a Mega. Can I use it or another 8266? Are there problems with voltages ? I mean 3,3v and 5v in wiring both.  Anyway the example refers to an 8266 8pins and mine is 30pins. I don´t know if it applies to my case

Than you for sharing your knowledge

**Reply by JAndrassy** [reply]
you have a nodemcu dev board, not a basic esp8266 module. just execute the esptool command 
```
esptool.py write_flash --flash_size 2MB-c1 0x0 boot_v1.7.bin 0x01000 at/1024+1024/user1.2048.new.5.bin 0x1fb000 blank.bin 0x1fc000 esp_init_data_default_v08.bin 0xfe000 blank.bin 0x1fe000 blank.bin
```
